swiftsimio/initial_conditions/generate_particles.py

```diff
@@ -1,5 +1,1304 @@
+
+        avg_displacement: float
+           average displacement of this iteration step.
+
+        """
+
+        self.max_displacement[iteration] = max_displacement
+        self.min_displacement[iteration] = min_displacement
+        self.avg_displacement[iteration] = avg_displacement
+        self.niter = iteration
+        return
+
+
+class ParticleGenerator(object):
+    r"""
+    Main class for generating initial conditions for SPH simulations for
+    a given density function ``rho(x, ndim)`` following Arth et al. 2019
+    (https://arxiv.org/abs/1907.11250).
+
+    Attributes
+    -------------
+
+    self.rhofunc: callable
+        The density function that is to be reproduced in the initial conditions.
+        It must take two positional arguments:
+        
+        - ``x``: np.ndarray of 3D particle coordinates (even if your initial 
+          conditions have lower dimensionality)
+        - ``ndim``: integer, number of dimensions that your simulations is to have
+
+    self.boxsize: unyt.unyt_array
+        The box size of the simulation.
+
+    self.nx: int
+        how many particles along every dimension you want your simulation
+        to contain
+
+    self.ndim: int
+        how many dimensions you want your simulation to have
+
+    self.unitsys: unyt.unit_systems.UnitSystem
+        a unit system that contains the units you want your ICs to have
+
+    self.periodic: bool
+        whether the simulation box is periodic or not
+
+    self.kernel: str {'cubic spline',}
+        which kernel to use
+
+    self.eta: float, optional
+        resolution eta, which defines the number of neighbours used
+        independently of dimensionality
+
+    self.rho_max: float or None
+        The maximal density within the simulation box. If ``None``, an 
+        approximate value will be determined if the rejection sampling to obtain
+        an initial particle configuration guess is used.
+
+    self.runparams: RunParams
+        A ``RunParams`` instance that contains runtime parameters for the generator.
+        The user is encouraged to tinker with these parameters before the iteration
+        is commenced.
+
+    self.iterparams: IterData
+        An ``IterData`` instance that holds data relevant for iteration steps. The
+        user is *strongly discouraged* of fiddling with this.
+
+    self.stats: IterStats
+        An ``IterStats`` instance that holds iteration statistics that should be
+        accessible after the iteration finished.
+
+    self.npart: int
+        total number of particles in simulation
+
+    self.coordinates: unyt.unyt_array
+        unyt_array containing the final particle coordinates with shape (self.npart, 3).
+
+    self.masses: unyt.unyt_array
+        unyt_array containing the final particle masses with shape (self.npart).
+
+    self.smoothing_lengths: unyt.unyt_array
+        unyt_array containing the final particle smoothing lengths with shape (self.npart, 3).
```

Comment:
Make sure that your comments don't exceed 88 characters as unfortunately `black` doesn't wrap them :(

Reply:
Is that really that necessary? As far as I see, sphinx doesn't really care about where I put linebreaks

Reply:
Did it anyway.